[cmake] Refactor DIA SDK detection into FindDIASDK module

ZhongRuoyu opened the pull request:
This consolidates the DIA SDK detection logic from `{llvm,compiler-rt}/cmake/config-ix.cmake` into a new centralized, reusable `FindDIASDK.cmake` module.

In addition to code deduplication, it also helps to avoid hard-coded references to the DIA SDK location in `LLVMExports.cmake`, hence allowing a pre-built LLVM distribution for Windows to be used on another host without requiring the DIA SDK location to be the same.

Fixes https://github.com/llvm/llvm-project/issues/86250.
Fixes https://github.com/llvm/llvm-project/issues/100372.
Fixes https://github.com/llvm/llvm-project/issues/111829.
Fixes https://github.com/llvm/llvm-project/issues/152268.

CC @petrhosek (LLVM CMake maintainer).

ZhongRuoyu committed:
[cmake] Refactor DIA SDK detection into FindDIASDK module

This consolidates the DIA SDK detection logic from
{llvm,compiler-rt}/cmake/config-ix.cmake into a new centralized,
reusable FindDIASDK.cmake module.

In addition to code deduplication, it also helps to avoid hard-coded
references to the DIA SDK location in LLVMExports.cmake, hence allowing
a pre-built LLVM distribution for Windows to be used on another host
without requiring the DIA SDK location to be the same.

Fixes https://github.com/llvm/llvm-project/issues/86250.
Fixes https://github.com/llvm/llvm-project/issues/100372.
Fixes https://github.com/llvm/llvm-project/issues/111829.
Fixes https://github.com/llvm/llvm-project/issues/152268.

Signed-off-by: Ruoyu Zhong <zhongruoyu@outlook.com>

compnerd commented on llvm/cmake/modules/FindDIASDK.cmake:
```diff
@@ -0,0 +1,74 @@
+  set(MSVC_DIA_SDK_DIR "${LLVM_WINSYSROOT}/DIA SDK" CACHE PATH
+      "Path to the DIA SDK")
+else()
+  set(MSVC_DIA_SDK_DIR "$ENV{VSINSTALLDIR}DIA SDK" CACHE PATH
```
I guess this is the original code, but if `VSINSTALLDIR` is not set in the environment, this will default to `DIA SDK`.

compnerd replied:
It is a weird state and if your search paths are setup properly, it could potentially still find the libraries. I'm not sure if its worth the effort to clean that up, its a corner case that is unlikely to be hit and this does maintain the status quo.

compnerd reviewed (approved):
This definitely is a good cleanup of the handling and using a custom Find module is moving towards the current best practices for CMake usage.

compnerd commented on llvm/cmake/modules/FindDIASDK.cmake:
```diff
@@ -0,0 +1,74 @@
+# Additionally, the following import target will be defined:
+#   DIASDK::Diaguids
+
+if(NOT WIN32)
+  set(DIASDK_FOUND FALSE)
+  return()
+endif()
```
I don't think that conditional on `MSVC` is correct, this is not dependent on the MSVC generator nor the MSVC compiler, but rather if we are building for Windows. That said, the guard here is just avoid the unnecessary stats, which is a nice touch.

ZhongRuoyu committed:
[cmake] Add diagnostic message for MSVC_DIA_SDK_DIR edge case

It is technically possible that MSVC_DIA_SDK_DIR could not be inferred
from LLVM_WINSYSROOT or VSINSTALLDIR. In that case, don't attempt to set
it to a broken path at all, and emit a diagnostic message instead.

Signed-off-by: Ruoyu Zhong <zhongruoyu@outlook.com>

compnerd commented on llvm/cmake/modules/FindDIASDK.cmake:
```diff
@@ -0,0 +1,77 @@
+  set(MSVC_DIA_SDK_DIR "$ENV{VSINSTALLDIR}DIA SDK" CACHE PATH
+      "Path to the DIA SDK")
+elseif(NOT DEFINED MSVC_DIA_SDK_DIR)
+  message(STATUS "MSVC_DIA_SDK_DIR not set, and could not be inferred. DIA SDK "
```
Should this use `SEND_WARNING`?

ZhongRuoyu replied:
IMO probably not, because the usage of `find(DIASDK)` may be speculative and I would defer to the caller to emit that warning if necessary.

ZhongRuoyu commented:
Gentle ping -- happy to address any additional concerns. Otherwise, it would be nice to land this and backport it to the 21.x release branch as this helps to fix a long-standing issue in the official distribution tarballs that affected a non-trivial number of users.

I don't have commit access so I'll need your help to merge this. I'd appreciate your endorsement when I fulfill the requirements for commit access application.

ZhongRuoyu committed:
Merge remote-tracking branch 'origin/main' into cmake-diaguids

ZhongRuoyu commented:
I'll merge this and backport it to 22.x given that it's still being reported (see latest comments in e.g. https://github.com/llvm/llvm-project/issues/86250).

ZhongRuoyu commented:
/cherry-pick 3545244333dfc64de756ad13f7e3792d79989898

ZhongRuoyu commented:
/cherry-pick 3545244333dfc64de756ad13f7e3792d79989898

mstorsjo commented:
FWIW, the use of the DIA SDK is being phased out as well, so this might not be relevant at all, soon.
